catalog/cataloger_export.go

```diff
@@ -139,107 +140,88 @@ func (c *cataloger) PutExportConfiguration(repository string, branch string, con
 
 var ErrExportFailed = errors.New("export failed")
 
-func (c *cataloger) ExportState(repo, branch, newRef string, cb func(oldRef string, state CatalogBranchExportStatus) (newState CatalogBranchExportStatus, newMessage *string, err error)) error {
-	_, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
-		oldRef, state, err := c.ExportStateMarkStart(tx, repo, branch, newRef)
-		if err != nil {
-			return nil, err
-		}
+type ExportState struct {
+	CurrentRef   string
+	State        CatalogBranchExportStatus
+	ErrorMessage *string
+}
 
-		newState, newMsg, err := cb(oldRef, state)
+func (c *cataloger) GetExportState(repo string, branch string) (ExportState, error) {
+	res, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+		var res ExportState
+
+		branchID, err := c.getBranchIDCache(tx, repo, branch)
 		if err != nil {
 			return nil, err
 		}
-		err = c.ExportStateMarkEnd(tx, repo, branch, newRef, newState, newMsg)
-		return nil, err
+		// get current state
+		err = tx.Get(&res, `
+		SELECT current_ref, state, error_message
+		FROM catalog_branches_export_state
+		WHERE branch_id=$1`,
+			branchID)
+		return res, err
 	})
-	return err
+	return res.(ExportState), err
 }
 
-func (c *cataloger) ExportStateMarkStart(tx db.Tx, repo string, branch string, newRef string) (oldRef string, state CatalogBranchExportStatus, err error) {
-	var res struct {
-		CurrentRef   string
-		State        CatalogBranchExportStatus
-		ErrorMessage *string
-	}
-	branchID, err := c.getBranchIDCache(tx, repo, branch)
-	if err != nil {
-		return oldRef, state, err
-	}
-	err = tx.Get(&res, `
+func (c *cataloger) ExportState(repo string, branch string, newRef string, cb stateCB) error {
+	_, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+		var res struct {
+			CurrentRef   string
+			State        CatalogBranchExportStatus
+			ErrorMessage *string
+		}
+
+		branchID, err := c.getBranchIDCache(tx, repo, branch)
+		if err != nil {
+			return nil, err
+		}
+		// get current state
+		err = tx.Get(&res, `
 		SELECT current_ref, state, error_message
 		FROM catalog_branches_export_state
 		WHERE branch_id=$1 FOR UPDATE`,
-		branchID)
-	missing := errors.Is(err, db.ErrNotFound)
-	if err != nil && !missing {
-		err = fmt.Errorf("ExportStateuMarkStart: failed to get existing state: %w", err)
-		return oldRef, state, err
-	}
-	oldRef = res.CurrentRef
-	state = res.State
+			branchID)
+		missing := errors.Is(err, db.ErrNotFound)
+		if err != nil && !missing {
+			err = fmt.Errorf("ExportStateuMarkStart: failed to get existing state: %w", err)
+			return nil, err
+		}
+		oldRef := res.CurrentRef
+		state := res.State
 
-	var query string
-	if missing {
-		query = `
+		// run callback
+		newState, newMsg, err := cb(oldRef, state)
+		if err != nil {
+			return nil, err
+		}
+		// in case newRef was not set leave oldRef
```

Comment:
Yeah, so probably best to return `newRef` from the callback as well.  Otherwise it's a bit tricky to understand what a call does.  (This was also an issue with my previous version, of course!)

Reply:
👍